File: hw02/Pet.java
package hw02;

import java.util.Date;
import java.util.List;

/**
 * Pet is an abstract class inherited from base Animal class
 */
public abstract class Pet extends Animal{
    private String name;
    private String breed;
    private List<String> vaccinations;
    private String furColor;
    private Date birthDate;

    /**
     * It is a constructor for Pet class
     * @param height            The height of animal
     * @param weight            The weight of animal
     * @param eyeColor          The eye color of animal
     * @param name              The name of animal
     * @param breed             The breed of animal
     * @param vaccinations      The vaccinations list of animal
     * @param furColor          The fur color of animal
     * @param birthDate         The birthdate of animal
     */
    public Pet(double height, double weight, String eyeColor, String name, String breed, List<String> vaccinations, String furColor, Date birthDate) {
        super(height, weight, eyeColor);
        this.name = name;
        this.breed = breed;
        this.vaccinations = vaccinations;
        this.furColor = furColor;
        this.birthDate = birthDate;
    }

    /**
     * Returns the animal name
     * @return this.name
     */
    public String getName() {
        return this.name;
    }

    /**
     * Returns the animal breed
     * @return this.breed
     */
    public String getBreed() {
        return this.breed;
    }

    /**
     * Returns the animal's vaccinations list
     * @return this.vaccinations
     */
    public List<String> getVaccinations() {
        return this.vaccinations;
    }

    /**
     * Returns the animal fur color
     * @return this.furColor
     */
    public String getFurColor() {
        return this.furColor;
    }

    /**
     * Returns Returns the animal birth date
     * @return this.birthDate
     */
    public Date getBirthDate() {
        return this.birthDate;
    }

    public void pet() {
        System.out.println("It has been petted. It loves you)");
    }

    @Override
    public String toString() {
        return String.format(
            "%s, Name: %s, Breed: %s, Vaccinations: %s, Fur color: %s, Birth date: %s",
            super.toString(), this.getName(), this.getBreed(), this.getVaccinations(), this.getFurColor(), this.getBirthDate()
        );
    }
}

File: hw07/internal/models/pet.go
package models

import (
	"fmt"
	"time"
)

// Pet base structure
type Pet struct {
	Animal
	name         string
	breed        string
	vaccinations []string
	furColor     string
	birthDate    time.Time
}

// Creates new Pet instance with full parameters given
func newPet(height float64, weight float64, eyeColor string, name string, breed string, vaccinations []string, furColor string, birthDate time.Time) (*Pet, error) {
	a, err := newAnimal(height, weight, eyeColor)
	if err != nil {
		return nil, err
	}

	p := &Pet{
		*a,
		name,
		breed,
		vaccinations,
		furColor,
		birthDate,
	}

	// Handles potential restrictions of any parameter to return error when new Pet cannot be instantiated
	return p, nil
}

// Returns the pet's name
func (p *Pet) Name() string {
	return p.name
}

// Returns the pet's breed
func (p *Pet) Breed() string {
	return p.breed
}

// Returns the pet's vaccinations list
func (p *Pet) Vaccinations() []string {
	return p.vaccinations
}

// Returns the pet's fur color
func (p *Pet) FurColor() string {
	return p.furColor
}

// Returns the pet's birth date
func (p *Pet) BirthDate() time.Time {
	return p.birthDate
}

// Pets the pet
func (p *Pet) PetMe() string {
	return fmt.Sprintf("%s has been petted. It loves you)", p.name)
}

// Returns the string representation of Pet instance
func (p *Pet) String() string {
	return fmt.Sprintf("%v, Name: %s, Breed: %s, Vaccinations: %s, Fur color: %s, Birth date: %s",
		p.Animal.String(), p.name, p.breed, p.vaccinations, p.furColor, p.birthDate.Format("02-01-2006"))
}
